src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -173,6 +172,290 @@ bool isTwoArgumentsFromDifferentSides(const FunctionNode & node_function, const
+
+    // There are at least two arguments in the passed-in OR expression, thus we either completely eliminated at least one arguments, or there should be at least 2 remaining arguments.
+    // The complete elimination is handled above, so at this point we can be sure there are at least 2 arguments.
+    chassert(new_or_arguments.size() >= 2);
+
+    auto new_or_node = std::make_shared<FunctionNode>("or");
+    new_or_node->getArguments().getNodes() = std::move(new_or_arguments);
+
+    auto or_function_resolver = FunctionFactory::instance().get("or", context);
+    new_or_node->resolveAsFunction(or_function_resolver);
+
+    return CommonExpressionExtractionResult{new_or_node, common_exprs};
+}
+
+void tryOptimizeCommonExpressionsInOr(QueryTreeNodePtr & node, const ContextPtr & context)
+{
+    auto * root_node = node->as<FunctionNode>();
+    chassert(root_node && root_node->getFunctionName() == "or");
+
+    if (auto maybe_result = tryExtractCommonExpressions(node, context); maybe_result.has_value())
+    {
+        auto & result = *maybe_result;
+        auto & root_arguments = root_node->getArguments().getNodes();
+        root_arguments = std::move(result.common_expressions);
+        if (result.new_node != nullptr)
+            root_arguments.push_back(std::move(result.new_node));
+
+        if (root_arguments.size() == 1)
+        {
+            node = std::move(root_arguments[0]);
+            return;
+        }
+
+        // The OR expression must be replaced by and AND expression that will contain the common expressions
+        // and the new_node, if it is not nullptr.
+        auto and_function_resolver = FunctionFactory::instance().get("and", context);
+        root_node->resolveAsFunction(and_function_resolver);
+    }
+}
+
+void tryOptimizeCommonExpressionsInAnd(QueryTreeNodePtr & node, const ContextPtr & context)
+{
+    auto * root_node = node->as<FunctionNode>();
+    chassert(root_node && root_node->getFunctionName() == "and");
+
+    QueryTreeNodePtrWithHashSet new_top_level_arguments_set;
+    QueryTreeNodes new_top_level_arguments;
+
+    auto insert_possible_new_top_level_arg = [&new_top_level_arguments_set, &new_top_level_arguments](QueryTreeNodePtr node_to_insert)
+    {
+        insertIfNotPresentInSet(new_top_level_arguments_set, new_top_level_arguments, std::move(node_to_insert));
+    };
+    auto extracted_something = false;
+    auto & root_arguments = root_node->getArguments();
+
+    for (const auto & argument : root_node->getArguments())
+    {
+        if (auto maybe_result = tryExtractCommonExpressions(argument, context); maybe_result.has_value())
+        {
+            extracted_something = true;
+            auto & result = *maybe_result;
+            if (result.new_node != nullptr)
+                insert_possible_new_top_level_arg(std::move(result.new_node));
+            for (auto& common_expr: result.common_expressions)
+                insert_possible_new_top_level_arg(std::move(common_expr));
```

Comment:
It seems that in the following case:
```
(A & B & C | A & B & D) & (B & A & E | B & A & F)
```
We will get the following result:
```
A & B & (C | D) & B & A & (E | F)
```

Because `A & B` is not the same in terms of query tree as `B & A`

Reply:
I checked this out and this is already handled correctly, the result will be:
`A & B & (C | D) & (E | F)` because the arguments are flattened and considered one by one. The only issue would be if we would have `A` or `B` duplicated in one of the clauses, but:
 1. We have an optimization to eliminate duplicates that is done in `enterImpl`, so in `leaveImp` we know there are no duplicates.
 2. I slightly changed the code [(here before my changes)](https://github.com/ClickHouse/ClickHouse/blob/13750c444a7b3406f022e27b8a9176180f7e856f/src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp#L313-L320) to make sure this won't happen. I will also add some tests to verify.

Reply:
> We have an optimization to eliminate duplicates

I was wrong here. That only works for equalities. But my changes ensure that all the extracted expressions are deduplicated. I added your case to the tests.